interfaces/builtin/unity8.go

```diff
@@ -0,0 +1,190 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016-2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+	"strings"
+	"unicode"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/snap"
+)
+
+var unity8ConnectedPlugAppArmor = []byte(`
+# Description: Can access unity8 desktop services
+
+  #include <abstractions/dbus-session-strict>
+
+  # Fonts
+  #include <abstractions/fonts>
+  /var/cache/fontconfig/   r,
+  /var/cache/fontconfig/** mr,
+
+  # Accessibility
+  #include <abstractions/dbus-accessibility-strict>
+  dbus (send)
+       bus=session
+       path=/org/a11y/bus
+       interface=org.a11y.Bus
+       member=GetAddress
+       peer=(name=org.a11y.Bus,label=unconfined),
+  dbus (receive, send)
+       bus=accessibility
+       path=/org/a11y/atspi/**
+       peer=(label=unconfined),
+
+  # URL dispatcher. All apps can call this since:
+  # a) the dispatched application is launched out of process and not
+  #    controllable except via the specified URL
+  # b) the list of url types is strictly controlled
+  # c) the dispatched application will launch in the foreground over the
+  #    confined app
+  dbus (send)
+       bus=session
+       path=/com/canonical/URLDispatcher
+       interface=com.canonical.URLDispatcher
+       member=DispatchURL
+       peer=(name=com.canonical.URLDispatcher,label=###SLOT_SECURITY_TAGS###),
+
+  # This is needed when the app is already running and needs to be passed in
+  # a URL to open. This is most often used with content-hub providers and
+  # url-dispatcher, but is actually supported by Qt generally.
+  dbus (receive)
+       bus=session
+       path=/###PLUG_DBUS_APPIDS###
+       interface=org.freedesktop.Application
+       member=Open
+       peer=(label=###SLOT_SECURITY_TAGS###),
+
+  # Unity launcher (e.g. app counter, progress, alert)
+  dbus (receive, send)
+       bus=session
+       path=/com/canonical/Unity/Launcher/###PLUG_DBUS_APPIDS###
+       peer=(name=com.canonical.Unity.Launcher,label=###SLOT_SECURITY_TAGS###),
+
+  # Content Hub (pasteboard, file transfers)
+  dbus (receive, send)
+       bus=session
+       interface=com.ubuntu.content.dbus.Service
+       path=/
+       peer=(name=com.ubuntu.content.dbus.Service,label=###SLOT_SECURITY_TAGS###),
```

Comment:
We have a lot more interesting policy on Touch for content transfer. I suspect content-hub will be a separate interface that unity8 will implement because, iirc, content-hub may want to be used elsewhere for file transfer. For now, can you adjust this to be something along the lines of:
```
# Note: content-hub may become its own interface, but for now include it here
# Pasteboard via Content Hub
dbus (receive, send)
     bus=session
     interface="com.ubuntu.content.dbus.Service"
     path="/"
     member="{CreatePaste,GetPasteData,GetLatestPasteData,PasteFormats,"
     peer=(name=com.ubuntu.content.dbus.Service,label=###SLOT_SECURITY_TAGS###),
dbus (receive)
     bus=session
     interface="com.ubuntu.content.dbus.Service"
     path="/"
     member="PasteFormatsChanged"
     peer=(name=com.ubuntu.content.dbus.Service,label=###SLOT_SECURITY_TAGS###),

# File transfer via Content Hub
dbus (send)
     bus=session
     interface=org.freedesktop.DBus
     path=/org/freedesktop/DBus
     member={Request,Release}Name
     peer=(label=unconfined),
dbus (bind)
     bus=session
     name=com.ubuntu.content.handler.@{PROFILE_DBUS},
dbus (receive)
     bus=session
     path=/com/ubuntu/content/handler/@{PROFILE_DBUS}
     interface=com.ubuntu.content.dbus.Handler
     peer=(label=###SLOT_SECURITY_TAGS###),

# File transfer source via Content Hub
dbus (receive, send)
     bus=session
     interface=com.ubuntu.content.dbus.Transfer
     path=/transfers/@{PROFILE_DBUS}/export/*
     peer=(label=###SLOT_SECURITY_TAGS###),

# File transfer destination via Content Hub
dbus (receive, send)
     bus=session
     interface=com.ubuntu.content.dbus.Transfer
     path=/transfers/@{PROFILE_DBUS}/import/*
     peer=(label=###SLOT_SECURITY_TAGS###),

# LP: #1293771
# Since fd delegation doesn't exist in the form that we need it at this time,
# content-hub will create hard links in 'HubIncoming' for volatile data. As
# such, snaps should not have write access to anything in this directory
# otherwise they would be able to change the source content.
audit deny @{HOME}/snap/@{SNAP_NAME}/**/HubIncoming/** w,
```

I used a glob for HubIncoming because I don't know where that is these days. It could be more specific, but we'd have to make sure that the content-hub implementation never changed it without changing this interface.

Lastly, the Touch policy also had this rule which covers all of the paste rules:
```
dbus (receive, send)
     bus=session
     interface=com.ubuntu.content.dbus.Service
     peer=(label=unconfined),
```

This is close to the rule you used, but I'd prefer to leave it out for now. Can you use the recommended policy and report back any issues (pasting the denials) so that we can work through additional policy?